### PR TITLE
Release 0.4.0

### PR DESCRIPTION
**Pull request by sunnycase**
## Major changes
- Add FPIOA configuration
- Add I2C clock rate setting
- Support variable cpu frequency
## Breaking changes
- AES SHA256 FFT API reworks
- Make drivers private by default
## Non-breaking bug fixes
- Fix the procedure of setting pll
- Fix bugs in syscalls
- Fix incomplete initialization of sbss
- Fix task critical in interrupt

### CODE REVIEW

**Comment by ly0** [comment]
LGTM

File: lib/arch/include/platform.h [hunk 1]
```diff
@@ -0,0 +1,807 @@
...
+    FUNC_SPI0_D7          = 11, /*!< SPI0 Data 7 */
+    FUNC_SPI0_SS0         = 12, /*!< SPI0 Chip Select 0 */
+    FUNC_SPI0_SS1         = 13, /*!< SPI0 Chip Select 1 */
+    FUNC_SPI0_SS2         = 14, /*!< SPI0 Chip Select 2 */
+    FUNC_SPI0_SS3         = 15, /*!< SPI0 Chip Select 3 */
+    FUNC_SPI0_ARB         = 16, /*!< SPI0 Arbitration */
+    FUNC_SPI0_SCLK        = 17, /*!< SPI0 Serial Clock */
+    FUNC_UARTHS_RX        = 18, /*!< UART High speed Receiver */
+    FUNC_UARTHS_TX        = 19, /*!< UART High speed Transmitter */
+    FUNC_CLK_IN1          = 20, /*!< Clock Input 1 */
```
Comment (vowstar) [inline lib/arch/include/platform.h]: Confirm this please @sky123450

File: lds/kendryte.ld [hunk 2]
```diff
@@ -161,6 +163,12 @@ SECTIONS
     *(.dynbss)
     *(.bss .bss.* .gnu.linkonce.b.*)
     *(COMMON)
+	/* Align here to ensure that the .bss section occupies space up to
+       _end.  Align after .bss to ensure correct alignment even if the
+       .bss section disappears because there are no input sections.
+       FIXME: Why do we need it? When there is no .bss section, we don't
+       pad the .data section.  */
+    . = ALIGN(. != 0 ? 64 / 8 : 1);
   } > ram : DYN_DATA
```
[the remark this answers is not in the document]
Comment (ly0) [reply lds/kendryte.ld]: 骚海星

File: lib/arch/include/platform.h [hunk 3]
```diff
@@ -0,0 +1,807 @@
...
+    FUNC_CLK_IN2          = 21, /*!< Clock Input 2 */
```
Comment (vowstar) [inline lib/arch/include/platform.h]: Confirm this please @sky123450

File: lib/bsp/device/gpiohs.c [hunk 4]
```diff
@@ -25,8 +25,8 @@
 #include "fpioa_cfg.h"
 
 #define COMMON_ENTRY_NO_PIN                                                        \
-    gpiohs_data* data = (gpiohs_data*)userdata;                                    \
-    volatile gpiohs_t* gpiohs = (volatile gpiohs_t*)data->base_addr; \
+    gpiohs_data *data = (gpiohs_data *)userdata;                                   \
+    volatile gpiohs_t *gpiohs = (volatile gpiohs_t *)data->base_addr;              \
     (void)gpiohs;
```
Comment (vowstar) [inline lib/bsp/device/gpiohs.c]: LGTM

File: lib/bsp/device/gpiohs.c [hunk 5]
```diff
@@ -72,22 +72,22 @@ static void gpiohs_install(void* userdata)
...
 {
 }
 
-static void gpiohs_set_drive_mode(uint32_t pin, gpio_drive_mode_t mode, void* userdata)
+static void gpiohs_set_drive_mode(uint32_t pin, gpio_drive_mode_t mode, void *userdata)
 {
     COMMON_ENTRY;
-    int io_number = fpioa_get_io_by_func(FUNC_GPIOHS0 + pin);
+    int io_number = fpioa_get_io_by_function(FUNC_GPIOHS0 + pin);
     configASSERT(io_number > 0);
```
Comment (vowstar) [inline lib/bsp/device/gpiohs.c]: LGTM

**Comment by vowstar** [comment]
LGTM

**Pull request merged**